// paymanai/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnprocessableEntityError = exports.PermissionDeniedError = exports.InternalServerError = exports.AuthenticationError = exports.BadRequestError = exports.RateLimitError = exports.ConflictError = exports.NotFoundError = exports.APIUserAbortError = exports.APIConnectionTimeoutError = exports.APIConnectionError = exports.APIError = exports.PaymanaiError = exports.fileFromPath = exports.toFile = exports.Paymanai = void 0;
const Core = __importStar(require("./core.js"));
const Errors = __importStar(require("./error.js"));
const Uploads = __importStar(require("./uploads.js"));
const API = __importStar(require("./resources/index.js"));
const balances_1 = require("./resources/balances.js");
const me_1 = require("./resources/me.js");
const payments_1 = require("./resources/payments.js");
const spend_limits_1 = require("./resources/spend-limits.js");
const version_1 = require("./resources/version.js");
/**
 * API Client for interfacing with the Paymanai API.
 */
class Paymanai extends Core.APIClient {
    /**
     * API Client for interfacing with the Paymanai API.
     *
     * @param {string | undefined} [opts.xPaymanAPISecret=process.env['PAYMAN_API_SECRET'] ?? undefined]
     * @param {string} [opts.baseURL=process.env['PAYMANAI_BASE_URL'] ?? https://agent.payman.ai/api] - Override the default base URL for the API.
     * @param {number} [opts.timeout=1 minute] - The maximum amount of time (in milliseconds) the client will wait for a response before timing out.
     * @param {number} [opts.httpAgent] - An HTTP agent used to manage HTTP(s) connections.
     * @param {Core.Fetch} [opts.fetch] - Specify a custom `fetch` function implementation.
     * @param {number} [opts.maxRetries=2] - The maximum number of times the client will retry a request.
     * @param {Core.Headers} opts.defaultHeaders - Default headers to include with every request to the API.
     * @param {Core.DefaultQuery} opts.defaultQuery - Default query parameters to include with every request to the API.
     */
    constructor({ baseURL = Core.readEnv('PAYMANAI_BASE_URL'), xPaymanAPISecret = Core.readEnv('PAYMAN_API_SECRET'), ...opts } = {}) {
        if (xPaymanAPISecret === undefined) {
            throw new Errors.PaymanaiError("The PAYMAN_API_SECRET environment variable is missing or empty; either provide it, or instantiate the Paymanai client with an xPaymanAPISecret option, like new Paymanai({ xPaymanAPISecret: 'My X Payman API Secret' }).");
        }
        const options = {
            xPaymanAPISecret,
            ...opts,
            baseURL: baseURL || `https://agent.payman.ai/api`,
        };
        super({
            baseURL: options.baseURL,
            timeout: options.timeout ?? 60000 /* 1 minute */,
            httpAgent: options.httpAgent,
            maxRetries: options.maxRetries,
            fetch: options.fetch,
        });
        this.version = new API.Version(this);
        this.me = new API.Me(this);
        this.balances = new API.Balances(this);
        this.payments = new API.Payments(this);
        this.spendLimits = new API.SpendLimits(this);
        this._options = options;
        this.xPaymanAPISecret = xPaymanAPISecret;
    }
    defaultQuery() {
        return this._options.defaultQuery;
    }
    defaultHeaders(opts) {
        return {
            ...super.defaultHeaders(opts),
            Accept: 'application/vnd.payman.v1+json',
            ...this._options.defaultHeaders,
        };
    }
    authHeaders(opts) {
        return { 'x-payman-api-secret': this.xPaymanAPISecret };
    }
}
exports.Paymanai = Paymanai;
_a = Paymanai;
Paymanai.Paymanai = _a;
Paymanai.DEFAULT_TIMEOUT = 60000; // 1 minute
Paymanai.PaymanaiError = Errors.PaymanaiError;
Paymanai.APIError = Errors.APIError;
Paymanai.APIConnectionError = Errors.APIConnectionError;
Paymanai.APIConnectionTimeoutError = Errors.APIConnectionTimeoutError;
Paymanai.APIUserAbortError = Errors.APIUserAbortError;
Paymanai.NotFoundError = Errors.NotFoundError;
Paymanai.ConflictError = Errors.ConflictError;
Paymanai.RateLimitError = Errors.RateLimitError;
Paymanai.BadRequestError = Errors.BadRequestError;
Paymanai.AuthenticationError = Errors.AuthenticationError;
Paymanai.InternalServerError = Errors.InternalServerError;
Paymanai.PermissionDeniedError = Errors.PermissionDeniedError;
Paymanai.UnprocessableEntityError = Errors.UnprocessableEntityError;
Paymanai.toFile = Uploads.toFile;
Paymanai.fileFromPath = Uploads.fileFromPath;
Paymanai.Version = version_1.Version;
Paymanai.Me = me_1.Me;
Paymanai.Balances = balances_1.Balances;
Paymanai.Payments = payments_1.Payments;
Paymanai.SpendLimits = spend_limits_1.SpendLimits;
var uploads_1 = require("./uploads.js");
Object.defineProperty(exports, "toFile", { enumerable: true, get: function () { return uploads_1.toFile; } });
Object.defineProperty(exports, "fileFromPath", { enumerable: true, get: function () { return uploads_1.fileFromPath; } });
var error_1 = require("./error.js");
Object.defineProperty(exports, "PaymanaiError", { enumerable: true, get: function () { return error_1.PaymanaiError; } });
Object.defineProperty(exports, "APIError", { enumerable: true, get: function () { return error_1.APIError; } });
Object.defineProperty(exports, "APIConnectionError", { enumerable: true, get: function () { return error_1.APIConnectionError; } });
Object.defineProperty(exports, "APIConnectionTimeoutError", { enumerable: true, get: function () { return error_1.APIConnectionTimeoutError; } });
Object.defineProperty(exports, "APIUserAbortError", { enumerable: true, get: function () { return error_1.APIUserAbortError; } });
Object.defineProperty(exports, "NotFoundError", { enumerable: true, get: function () { return error_1.NotFoundError; } });
Object.defineProperty(exports, "ConflictError", { enumerable: true, get: function () { return error_1.ConflictError; } });
Object.defineProperty(exports, "RateLimitError", { enumerable: true, get: function () { return error_1.RateLimitError; } });
Object.defineProperty(exports, "BadRequestError", { enumerable: true, get: function () { return error_1.BadRequestError; } });
Object.defineProperty(exports, "AuthenticationError", { enumerable: true, get: function () { return error_1.AuthenticationError; } });
Object.defineProperty(exports, "InternalServerError", { enumerable: true, get: function () { return error_1.InternalServerError; } });
Object.defineProperty(exports, "PermissionDeniedError", { enumerable: true, get: function () { return error_1.PermissionDeniedError; } });
Object.defineProperty(exports, "UnprocessableEntityError", { enumerable: true, get: function () { return error_1.UnprocessableEntityError; } });
exports = module.exports = Paymanai;
exports.default = Paymanai;
