// paymanai/resources/payments.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../resource.mjs";
import { isRequestOptions } from "../core.mjs";
export class Payments extends APIResource {
    /**
     * Create a new payee (aka payee) for future payments to be sent to
     */
    createPayee(body, options) {
        return this._client.post('/payments/payees', {
            body,
            ...options,
            headers: { Accept: 'application/vnd.payman.v1+json', ...options?.headers },
        });
    }
    /**
     * Delete a payee (aka payee)
     */
    deletePayee(id, options) {
        return this._client.delete(`/payments/payees/${id}`, {
            ...options,
            headers: { Accept: 'application/vnd.payman.v1+json', ...options?.headers },
        });
    }
    searchPayees(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.searchPayees({}, query);
        }
        return this._client.get('/payments/search-payees', {
            query,
            ...options,
            headers: { Accept: 'application/vnd.payman.v1+json', ...options?.headers },
        });
    }
    /**
     * Sends funds from an agent controlled wallet to a payee.
     */
    sendPayment(body, options) {
        return this._client.post('/payments/send-payment', {
            body,
            ...options,
            headers: {
                Accept: 'application/vnd.payman.v1+json',
                ...options?.headers,
            },
        });
    }
}
