// paymanai/resources/payments.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Payments = void 0;
const resource_1 = require("../resource.js");
const core_1 = require("../core.js");
class Payments extends resource_1.APIResource {
    /**
     * Create a new payee (aka payee) for future payments to be sent to
     */
    createPayee(body, options) {
        return this._client.post('/payments/payees', {
            body,
            ...options,
            headers: { Accept: 'application/vnd.payman.v1+json', ...options?.headers },
        });
    }
    /**
     * Delete a payee (aka payee)
     */
    deletePayee(id, options) {
        return this._client.delete(`/payments/payees/${id}`, {
            ...options,
            headers: { Accept: 'application/vnd.payman.v1+json', ...options?.headers },
        });
    }
    searchPayees(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.searchPayees({}, query);
        }
        return this._client.get('/payments/search-payees', {
            query,
            ...options,
            headers: { Accept: 'application/vnd.payman.v1+json', ...options?.headers },
        });
    }
    /**
     * Sends funds from an agent controlled wallet to a payee.
     */
    sendPayment(body, options) {
        return this._client.post('/payments/send-payment', {
            body,
            ...options,
            headers: {
                Accept: 'application/vnd.payman.v1+json',
                ...options?.headers,
            },
        });
    }
}
exports.Payments = Payments;
